File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.js
const minRemoveToMakeValid = (s) => {
  const st = [''];
  let p = 0, u = -1;
  for (const char of s) {
    if (char === '(') {
      st[p] += '(';
      st.push('');
      u = p++;
    }
    else if (char === ')') {
      if (st[u]) {
        st[p--] += ')';
        st[u--] += st.pop();
      }
    }
    else { st[p] += char; }
  }
  while (u > -1) {
    st[u] = st[u].slice(0,-1) + st.pop();
    u--;
  }
  return st[0];
};

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.py
class Solution:
  def minRemoveToMakeValid(self, s: str) -> str:
    def fix(s, op, cl):
      ans = ''
      lv = 0
      for ch in s:
        if ch == op:
          lv += 1
        elif ch == cl:
          if lv:
            lv -= 1    
          else:
            continue
        ans += ch
      return ans
    return fix(fix(s,'(',')')[::-1],')','(')[::-1]